### ISSUE TITLE
spark runtime attributes in WDL task

### ISSUE THREAD

**Issue by blueskypie** [broadinstitute/cromwell#2640]
Currently the spark runtime block only have [four attributes](https://github.com/broadinstitute/cromwell#spark-runtime-attributes). But the `gatk-launch` actually produces many attributes for `spark-submit` command, e.g.
```
--conf spark.driver.userClassPathFirst=true 
--conf spark.io.compression.codec=lzf 
--conf 'spark.executor.extraJavaOptions=-DGATK_STACKTRACE_ON_USER_EXCEPTION=true -Dsnappy.disable=true' 
...
```
And there are many attributes users may add, e.g. `--driver-cores` and `--driver-memory`.

I wonder if the runtime block can be changed to something like this
```
runtime {
  --class org.broadinstitute.hellbender.Main 
  --driver-cores 4 
  --driver-memory 8g 
  --total-executor-cores 10  
  --executor-memory 4g 
  --conf spark.io.compression.codec=lzf
--conf 'spark.executor.extraJavaOptions=-DGATK_STACKTRACE_ON_USER_EXCEPTION=true -Dsnappy.disable=true' 
  ...
        }
```
So users can put any options into it and those strings are added to spark-submit command w/o any change.

**Comment by jainh** [comment]
@blueskypie  is it required to repeat `--conf` ? also are you expecting `--conf` value to be in single quote ?

**Comment by jainh** [comment]
@geoffjentry something like this ` 'key=value key1=value1' ` can be received as runtime value for key `--conf`  ?

**Comment by blueskypie** [comment]
@jainh Yes, there are multiple `--conf` attributes.
Currently the `gatk-launch` does not produce the single quote; but if I run the following without single quote, it leads to error:
>Error: Unrecognized option: -Dsamjdk.use_async_io_read_samtools=false
command:
```
/opt/spark-latest/bin/spark-submit --master spark://localhost:6066 --deploy-mode cluster \
 --driver-cores 4 --driver-memory 8g --executor-memory 4g --total-executor-cores 10 \
 --conf 'spark.executor.extraJavaOptions=-DGATK_STACKTRACE_ON_USER_EXCEPTION=true -Dsamjdk.use_async_io_read_samtools=false'  \
--conf 'spark.driver.extraJavaOptions=-DGATK_STACKTRACE_ON_USER_EXCEPTION=true -Dsamjdk.use_async_io_read_samtools=false -Dsamjdk.use_async_io_write_samtools=false -Dsamjdk.use_async_io_write_tribble=false -Dsamjdk.compression_level=1 -Dsnappy.disable=true' \
...
```

**Comment by jainh** [comment]
@cjllanwarne @mcovarr  do you know if something like this is possible through runtime attributes?

**Comment by geoffjentry** [comment]
I believe a runtime attr can be an array if that helps

**Comment by jainh** [comment]
@geoffjentry so instead of repeating `--conf` we should pass it as an array ?

**Comment by geoffjentry** [comment]
@jainh what I'm saying is that I believe you could have a `conf` runtime attribute which takes an array of string (there are other array runtime attrs) and then for each element turn that into ` --conf ` argument. 

If I'm right that that's possible that's also assuming a) this works for the spark-y type of stuff at all and b) conf makes sense as a runtime attribute regardless

**Comment by blueskypie** [comment]
Just wonder why the runtime block has to be key-value pairs? `spark-submit` has tons of attributes and new attributes may be added in the future.  Can the runtime block just be wrapped as string and passed to `spark-submit`?

If it has to be key-value pair, can the key be something like "additionalArgs" and the value be a string of containing attributes the user wants to add? for example:
`"additionalArgs": "--conf 'xx -Dxx' --name xx"`
In this way, if `spark-submit` has new attributes in the future, cromwell doesn't need to updated.

**Comment by jainh** [comment]
@blueskypie  I don't know if we pass value in runtime section as this string `"additionalArgs": "--conf 'xx -Dxx' --name xx"` would be a legitimate construct , and will be parsed appropriately, @geoffjentry  correct me if i am wrong ? what @blueskypie  is asking is possible?

**Comment by kshakir** [comment]
Hi @jainh,

The runtime section is very backend-implementation specific, but to answer your question in the abstract:

The wdl spec [says](https://github.com/broadinstitute/wdl/blob/develop/SPEC.md#runtime-section) that "Values can be any expression …". For example:

Valid wdl string:
```
runtime {
  my_key: "a 'b c' d"
}
```

Valid wdl array:
```
runtime {
  my_key: ["a", "b c", "d"]
}
```

The following however is **invalid** according to the spec as the keys are duplicated:
```
runtime {
  my_key: "a"
  my_key: "b c"
  my_key: "d"
}
```

It is then up to the backend to decide and implement what keys and values it will accept.

The config backend is currently implemented to [only](https://github.com/broadinstitute/cromwell/blob/839ea1e456b929d6149430f4d7d3805f8c235d3f/supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/DeclarationValidation.scala#L43-L50) support primitive wdl types (WdlInteger, WdlString, WdlFloat, WdlBoolean) and their optional wrappers. While it does not support them, one could update that code to support arrays of values too.

Meanwhile, the JES backend already does support arrays for some attributes, e.g. for [`zones`](https://github.com/broadinstitute/cromwell/blob/839ea1e456b929d6149430f4d7d3805f8c235d3f/supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesRuntimeAttributes.scala#L127) and [`disks`](https://github.com/broadinstitute/cromwell/blob/839ea1e456b929d6149430f4d7d3805f8c235d3f/supportedBackends/jes/src/main/scala/cromwell/backend/impl/jes/JesRuntimeAttributes.scala#L142).

**Comment by jainh** [comment]
@kshakir thank you!

**Comment by jainh** [comment]
@blueskypie create this [PR](https://github.com/broadinstitute/cromwell/pull/2672) , any suggestion ?

**Comment by geoffjentry** [comment]
Closed by #2672
